File: rust/crates/oxa-chatcompletions/tests/stream.rs
use oxa_chatcompletions::{
    ChoiceDelta, Chunk, CompletionTokensDetailsWire, Config, DeltaPayload, PromptTokensDetailsWire,
    StreamDecoder, StreamEncoder, ToolCallDelta, UsageWire,
};
use oxa_ir::{
    Block, Delta, Event, InputTokensDetails, LossReason, OutputTokensDetails, StopReason, Usage,
};

fn chunk_role(id: &str, model: &str) -> Chunk {
    Chunk {
        id: id.to_string(),
        object: "chat.completion.chunk".to_string(),
        model: model.to_string(),
        choices: vec![ChoiceDelta {
            index: 0,
            delta: DeltaPayload {
                role: "assistant".to_string(),
                ..Default::default()
            },
            finish_reason: None,
        }],
        ..Default::default()
    }
}

fn chunk_content(s: &str) -> Chunk {
    Chunk {
        object: "chat.completion.chunk".to_string(),
        choices: vec![ChoiceDelta {
            index: 0,
            delta: DeltaPayload {
                content: Some(s.to_string()),
                ..Default::default()
            },
            finish_reason: None,
        }],
        ..Default::default()
    }
}

fn chunk_finish(reason: &str) -> Chunk {
    Chunk {
        object: "chat.completion.chunk".to_string(),
        choices: vec![ChoiceDelta {
            index: 0,
            delta: DeltaPayload::default(),
            finish_reason: Some(reason.to_string()),
        }],
        ..Default::default()
    }
}

fn chunk_usage() -> Chunk {
    Chunk {
        object: "chat.completion.chunk".to_string(),
        choices: Vec::new(),
        usage: Some(UsageWire {
            prompt_tokens: 3,
            completion_tokens: 5,
            total_tokens: 8,
            ..UsageWire::default()
        }),
        ..Default::default()
    }
}

#[test]
fn stream_decoder_maps_reasoning_before_text_and_keeps_usage_details() {
    let mut decoder = StreamDecoder::new(&Config::default());
    let mut got = Vec::new();
    got.extend(decoder.feed(&chunk_role("chatcmpl-m9", "o3-mini")).unwrap());
    got.extend(
        decoder
            .feed(&Chunk {
                id: "chatcmpl-m9".to_string(),
                model: "o3-mini".to_string(),
                choices: vec![ChoiceDelta {
                    index: 0,
                    delta: DeltaPayload {
                        reasoning_content: Some("Plan.".to_string()),
                        ..Default::default()
                    },
                    finish_reason: None,
                }],
                ..Default::default()
            })
            .unwrap(),
    );
    got.extend(
        decoder
            .feed(&Chunk {
                id: "chatcmpl-m9".to_string(),
                model: "o3-mini".to_string(),
                choices: vec![ChoiceDelta {
                    index: 0,
                    delta: DeltaPayload {
                        content: Some("Answer.".to_string()),
                        ..Default::default()
                    },
                    finish_reason: Some("stop".to_string()),
                }],
                usage: Some(UsageWire {
                    prompt_tokens: 3,
                    completion_tokens: 5,
                    total_tokens: 8,
                    prompt_tokens_details: Some(PromptTokensDetailsWire { cached_tokens: 0 }),
                    completion_tokens_details: Some(CompletionTokensDetailsWire {
                        reasoning_tokens: 2,
                    }),
                }),
                ..Default::default()
            })
            .unwrap(),
    );
    got.extend(decoder.flush().unwrap());

    assert_eq!(
        got,
        vec![
            Event::MessageStart {
                id: "chatcmpl-m9".to_string(),
                model: "o3-mini".to_string(),
            },
            Event::ContentBlockStart {
                index: 0,
                block: Block::Thinking {
                    thinking: String::new(),
                    signature: None,
                },
            },
            Event::ContentBlockDelta {
                index: 0,
                delta: Delta::ThinkingDelta {
                    text: "Plan.".to_string(),
                },
            },
            Event::ContentBlockStop { index: 0 },
            Event::ContentBlockStart {
                index: 1,
                block: Block::Text {
                    text: String::new(),
                },
            },
            Event::ContentBlockDelta {
                index: 1,
                delta: Delta::TextDelta {
                    text: "Answer.".to_string(),
                },
            },
            Event::ContentBlockStop { index: 1 },
            Event::MessageDelta {
                stop_reason: StopReason::EndTurn,
                stop_sequence: None,
                usage: Usage {
                    input_tokens: 3,
                    output_tokens: 5,
                    input_tokens_details: Some(InputTokensDetails { cached_tokens: 0 }),
                    output_tokens_details: Some(OutputTokensDetails {
                        reasoning_tokens: 2
                    }),
                    ..Usage::default()
                },
            },
            Event::MessageDone {},
        ]
    );
}

#[test]
fn stream_encoder_maps_thinking_and_usage_details() {
    let mut encoder = StreamEncoder::new(&Config::default());
    encoder
        .apply(&Event::MessageStart {
            id: "chatcmpl-m9".to_string(),
            model: "o3-mini".to_string(),
        })
        .unwrap();
    let (started, start_losses) = encoder
        .apply(&Event::ContentBlockStart {
            index: 0,
            block: Block::Thinking {
                thinking: String::new(),
                signature: Some("start-sig".to_string()),
            },
        })
        .unwrap();
    assert!(started.is_empty());
    assert_eq!(start_losses.len(), 1);
    let (thinking, losses) = encoder
        .apply(&Event::ContentBlockDelta {
            index: 0,
            delta: Delta::ThinkingDelta {
                text: "Plan.".to_string(),
            },
        })
        .unwrap();
    assert_eq!(
        thinking[0].choices[0].delta.reasoning_content.as_deref(),
        Some("Plan.")
    );
    assert!(losses.is_empty());
    let (signature, losses) = encoder
        .apply(&Event::ContentBlockDelta {
            index: 0,
            delta: Delta::SignatureDelta {
                signature: "delta-sig".to_string(),
            },
        })
        .unwrap();
    assert!(signature.is_empty());
    assert_eq!(losses.len(), 1);
    encoder
        .apply(&Event::ContentBlockStop { index: 0 })
        .unwrap();
    let (terminal, losses) = encoder
        .apply(&Event::MessageDelta {
            stop_reason: StopReason::EndTurn,
            stop_sequence: None,
            usage: Usage {
                input_tokens: 3,
                output_tokens: 5,
                input_tokens_details: Some(InputTokensDetails { cached_tokens: 0 }),
                output_tokens_details: Some(OutputTokensDetails {
                    reasoning_tokens: 2,
                }),
                ..Usage::default()
            },
        })
        .unwrap();
    assert!(losses.is_empty());
    let usage = terminal[0].usage.as_ref().unwrap();
    assert_eq!(
        usage.prompt_tokens_details.as_ref().unwrap().cached_tokens,
        0
    );
    assert_eq!(
        usage
            .completion_tokens_details
            .as_ref()
            .unwrap()
            .reasoning_tokens,
        2
    );
}

#[test]
fn stream_encoder_rejects_thinking_after_signature_delta() {
    let mut encoder = StreamEncoder::new(&Config::default());
    encoder
        .apply(&Event::MessageStart {
            id: "chatcmpl_late".to_string(),
            model: "o3-mini".to_string(),
        })
        .unwrap();
    encoder
        .apply(&Event::ContentBlockStart {
            index: 0,
            block: Block::Thinking {
                thinking: String::new(),
                signature: None,
            },
        })
        .unwrap();
    encoder
        .apply(&Event::ContentBlockDelta {
            index: 0,
            delta: Delta::SignatureDelta {
                signature: "sig".to_string(),
            },
        })
        .unwrap();

    let err = encoder
        .apply(&Event::ContentBlockDelta {
            index: 0,
            delta: Delta::ThinkingDelta {
                text: "late".to_string(),
            },
        })
        .expect_err("thinking cannot follow a signature");
    assert!(
        err.to_string()
            .contains("thinking delta after signature_delta")
    );
}

#[test]
fn stream_encoder_rejects_duplicate_signature_deltas() {
    let mut encoder = StreamEncoder::new(&Config::default());
    encoder
        .apply(&Event::MessageStart {
            id: "chatcmpl_duplicate".to_string(),
            model: "o3-mini".to_string(),
        })
        .unwrap();
    encoder
        .apply(&Event::ContentBlockStart {
            index: 0,
            block: Block::Thinking {
                thinking: String::new(),
                signature: None,
            },
        })
        .unwrap();
    encoder
        .apply(&Event::ContentBlockDelta {
            index: 0,
            delta: Delta::SignatureDelta {
                signature: "sig-1".to_string(),
            },
        })
        .unwrap();

    let err = encoder
        .apply(&Event::ContentBlockDelta {
            index: 0,
            delta: Delta::SignatureDelta {
                signature: "sig-2".to_string(),
            },
        })
        .expect_err("a thinking block permits one signature delta");
    assert!(err.to_string().contains("duplicate signature_delta"));
}

#[test]
fn stream_decoder_happy_path() {
    let config = Config::default();
    let mut d = StreamDecoder::new(&config);
    let mut got = Vec::new();

    got.extend(d.feed(&chunk_role("chatcmpl-1", "gpt-4o")).unwrap());
    got.extend(d.feed(&chunk_content("Hel")).unwrap());
    got.extend(d.feed(&chunk_content("lo")).unwrap());
    got.extend(d.feed(&chunk_finish("stop")).unwrap());
    got.extend(d.feed(&chunk_usage()).unwrap());
    got.extend(d.flush().unwrap());

    let want = vec![
        Event::MessageStart {
            id: "chatcmpl-1".to_string(),
            model: "gpt-4o".to_string(),
        },
        Event::ContentBlockStart {
            index: 0,
            block: Block::Text {
                text: String::new(),
            },
        },
        Event::ContentBlockDelta {
            index: 0,
            delta: Delta::TextDelta {
                text: "Hel".to_string(),
            },
        },
        Event::ContentBlockDelta {
            index: 0,
            delta: Delta::TextDelta {
                text: "lo".to_string(),
            },
        },
        Event::ContentBlockStop { index: 0 },
        Event::MessageDelta {
            stop_reason: StopReason::EndTurn,
            stop_sequence: None,
            usage: Usage {
                input_tokens: 3,
                output_tokens: 5,
                ..Usage::default()
            },
        },
        Event::MessageDone {},
    ];

    assert_eq!(got, want);
    assert!(d.losses().is_empty());
}

#[test]
fn stream_decoder_errors() {
    let config = Config::default();

    // Flush without finish_reason is an error
    let mut d = StreamDecoder::new(&config);
    d.feed(&chunk_role("c1", "m")).unwrap();
    assert!(d.flush().is_err());

    // Flush twice is an error
    let mut d = StreamDecoder::new(&config);
    d.feed(&chunk_finish("stop")).unwrap();
    assert!(d.flush().is_ok());
    assert!(d.flush().is_err());

    // Feed after flush is an error
    assert!(d.feed(&chunk_content("x")).is_err());

    // Duplicate finish_reason is an error
    let mut d = StreamDecoder::new(&config);
    d.feed(&chunk_finish("stop")).unwrap();
    assert!(d.feed(&chunk_finish("stop")).is_err());

    // Chunk stream already started with role
    let mut d = StreamDecoder::new(&config);
    d.feed(&chunk_role("c1", "m")).unwrap();
    assert!(d.feed(&chunk_role("c1", "m")).is_err());

    // Tool call non-consecutive index
    let mut d = StreamDecoder::new(&config);
    let bad_call = Chunk {
        choices: vec![ChoiceDelta {
            index: 0,
            delta: DeltaPayload {
                tool_calls: Some(vec![ToolCallDelta {
                    index: 5, // skip index 0
                    id: Some("call_1".to_string()),
                    kind: Some("function".to_string()),
                    function: None,
                }]),
                ..Default::default()
            },
            finish_reason: None,
        }],
        ..Default::default()
    };
    assert!(d.feed(&bad_call).is_err());
}

#[test]
fn stream_encoder_happy_path() {
    let config = Config::default();
    let mut e = StreamEncoder::new(&config);

    let (c1, l1) = e
        .apply(&Event::MessageStart {
            id: "chatcmpl-1".to_string(),
            model: "gpt-4o".to_string(),
        })
        .unwrap();
    assert_eq!(c1.len(), 1);
    assert_eq!(c1[0].choices[0].delta.role, "assistant");
    assert!(l1.is_empty());

    let (c2, _) = e
        .apply(&Event::ContentBlockStart {
            index: 0,
            block: Block::Text {
                text: String::new(),
            },
        })
        .unwrap();
    assert!(c2.is_empty());

    let (c3, _) = e
        .apply(&Event::ContentBlockDelta {
            index: 0,
            delta: Delta::TextDelta {
                text: "Hi".to_string(),
            },
        })
        .unwrap();
    assert_eq!(c3.len(), 1);
    assert_eq!(c3[0].choices[0].delta.content.as_deref(), Some("Hi"));

    let (c4, _) = e.apply(&Event::ContentBlockStop { index: 0 }).unwrap();
    assert!(c4.is_empty());

    let (c5, l5) = e
        .apply(&Event::MessageDelta {
            stop_reason: StopReason::EndTurn,
            stop_sequence: None,
            usage: Usage {
                input_tokens: 10,
                output_tokens: 2,
                ..Usage::default()
            },
        })
        .unwrap();
    assert_eq!(c5.len(), 1);
    assert_eq!(c5[0].choices[0].finish_reason.as_deref(), Some("stop"));
    assert_eq!(c5[0].usage.as_ref().map(|u| u.prompt_tokens), Some(10));
    assert!(l5.is_empty());

    let (c6, _) = e.apply(&Event::MessageDone {}).unwrap();
    assert!(c6.is_empty());
}

#[test]
fn stream_encoder_ordering_degrade_loss() {
    let config = Config::default();
    let mut e = StreamEncoder::new(&config);

    e.apply(&Event::MessageStart {
        id: "c1".to_string(),
        model: "m".to_string(),
    })
    .unwrap();

    // Tool block first
    e.apply(&Event::ContentBlockStart {
        index: 0,
        block: Block::ToolUse {
            id: "call_1".to_string(),
            name: "weather".to_string(),
            input: "{}".to_string(),
        },
    })
    .unwrap();
    e.apply(&Event::ContentBlockStop { index: 0 }).unwrap();

    // Text block after tool block -> ordering degradation
    e.apply(&Event::ContentBlockStart {
        index: 1,
        block: Block::Text {
            text: String::new(),
        },
    })
    .unwrap();
    e.apply(&Event::ContentBlockStop { index: 1 }).unwrap();

    let (_, losses) = e
        .apply(&Event::MessageDelta {
            stop_reason: StopReason::EndTurn,
            stop_sequence: None,
            usage: Usage {
                input_tokens: 0,
                output_tokens: 0,
                ..Usage::default()
            },
        })
        .unwrap();
    assert_eq!(losses.len(), 1);
    assert_eq!(losses[0].reason, LossReason::Degraded);
}

File: ts/src/openai/chatcompletions/stream.ts
import { OxaError } from "../../error.js";
import {
  encodeUsageInteger,
  parseUsageInteger,
  type Event,
  type StopReason,
  type Usage,
} from "../../ir/index.js";
import { jsonText, type JsonNumber } from "../../json/index.js";
import type { ConversionResult, Loss } from "../../loss.js";
import { mapModel, type ModelMapper } from "../../modelmap.js";

export interface ChatCompletionsChunk {
  readonly object?: "chat.completion.chunk";
  readonly created?: number;
  readonly id: string;
  readonly model: string;
  readonly choices: readonly ChatCompletionsChoice[];
  readonly usage?: ChatCompletionsUsage;
}
export interface ChatCompletionsChoice {
  readonly index?: number;
  readonly delta: ChatCompletionsDelta;
  readonly finish_reason: string | null;
}
export interface ChatCompletionsDelta {
  readonly role?: string;
  readonly content?: string;
  readonly reasoning_content?: string;
  readonly tool_calls?: readonly ChatCompletionsToolCallDelta[];
}
export interface ChatCompletionsToolCallDelta {
  readonly index: number;
  readonly id?: string;
  readonly type?: string;
  readonly function?: { readonly name?: string; readonly arguments?: string };
}
export interface ChatCompletionsUsage {
  readonly prompt_tokens: bigint | JsonNumber;
  readonly completion_tokens: bigint | JsonNumber;
  readonly total_tokens: bigint | JsonNumber;
  readonly prompt_tokens_details?: {
    readonly cached_tokens: bigint | JsonNumber;
  };
  readonly completion_tokens_details?: {
    readonly reasoning_tokens: bigint | JsonNumber;
  };
}
export interface ChatCompletionsStreamDecoderOptions {
  readonly modelMapper?: ModelMapper;
}
interface ToolCall {
  readonly nativeIndex: number;
  id: string | undefined;
  name: string;
  readonly fragments: string[];
  skipped: boolean;
}

/** Incrementally converts Chat Completions chunks to IR stream events. */
export class ChatCompletionsStreamDecoder {
  readonly #modelMapper: ModelMapper | undefined;
  readonly #losses: Loss[] = [];
  readonly #toolCalls: ToolCall[] = [];
  #started = false;
  #id = "";
  #nativeModel = "";
  #flushed = false;
  #finishSeen = false;
  #textOpen = false;
  #textIndex = 0;
  #thinkingOpen = false;
  #thinkingIndex = 0;
  #nextIrIndex = 0;
  #stopReason: StopReason = "other";
  #usage: Usage = { input_tokens: 0n, output_tokens: 0n };

  constructor(options: ChatCompletionsStreamDecoderOptions = {}) {
    this.#modelMapper = options.modelMapper;
  }

  Feed(chunk: ChatCompletionsChunk): readonly Event[] {
    if (this.#flushed) this.#lifecycle("chunk fed after stream flush");
    if (
      this.#started &&
      (chunk.id !== this.#id || chunk.model !== this.#nativeModel)
    )
      this.#lifecycle("chunk has a conflicting stream identity");
    if (chunk.usage !== undefined)
      this.#usage = {
        input_tokens: parseUsageInteger(
          chunk.usage.prompt_tokens,
          "usage.prompt_tokens",
        ),
        output_tokens: parseUsageInteger(
          chunk.usage.completion_tokens,
          "usage.completion_tokens",
        ),
        ...(chunk.usage.prompt_tokens_details === undefined
          ? {}
          : {
              input_tokens_details: {
                cached_tokens: parseUsageInteger(
                  chunk.usage.prompt_tokens_details.cached_tokens,
                  "usage.prompt_tokens_details.cached_tokens",
                ),
              },
            }),
        ...(chunk.usage.completion_tokens_details === undefined
          ? {}
          : {
              output_tokens_details: {
                reasoning_tokens: parseUsageInteger(
                  chunk.usage.completion_tokens_details.reasoning_tokens,
                  "usage.completion_tokens_details.reasoning_tokens",
                ),
              },
            }),
      };
    const choice = chunk.choices[0];
    if (choice === undefined) return [];
    if (this.#started && choice.delta.role !== undefined)
      this.#lifecycle(
        this.#finishSeen
          ? "chunk stream restarted after finish_reason"
          : "chunk stream already started",
      );
    const events: Event[] = [];
    if (!this.#started) {
      this.#started = true;
      this.#id = chunk.id;
      this.#nativeModel = chunk.model;
      events.push({
        type: "message_start",
        id: chunk.id,
        model: mapModel(this.#modelMapper, chunk.model),
      });
    }
    this.#recordToolCalls(choice.delta.tool_calls ?? []);
    if (choice.delta.reasoning_content !== undefined) {
      if (this.#textOpen) {
        events.push({ type: "content_block_stop", index: this.#textIndex });
        this.#textOpen = false;
      }
      if (!this.#thinkingOpen) {
        this.#thinkingOpen = true;
        this.#thinkingIndex = this.#nextIrIndex;
        this.#nextIrIndex += 1;
        events.push({
          type: "content_block_start",
          index: this.#thinkingIndex,
          block: { type: "thinking", thinking: "" },
        });
      }
      if (choice.delta.reasoning_content !== "")
        events.push({
          type: "content_block_delta",
          index: this.#thinkingIndex,
          delta: {
            type: "thinking_delta",
            text: choice.delta.reasoning_content,
          },
        });
    }
    if (choice.delta.content !== undefined) {
      if (this.#thinkingOpen) {
        events.push({ type: "content_block_stop", index: this.#thinkingIndex });
        this.#thinkingOpen = false;
      }
      if (!this.#textOpen) {
        this.#textOpen = true;
        this.#textIndex = this.#nextIrIndex;
        this.#nextIrIndex += 1;
        events.push({
          type: "content_block_start",
          index: this.#textIndex,
          block: { type: "text", text: "" },
        });
      }
      events.push({
        type: "content_block_delta",
        index: this.#textIndex,
        delta: { type: "text_delta", text: choice.delta.content },
      });
    }
    if (choice.finish_reason !== null) {
      if (this.#finishSeen) this.#lifecycle("duplicate finish_reason");
      this.#stopReason = this.#finish(choice.finish_reason);
      this.#finishSeen = true;
    }
    return events;
  }

  Flush(): readonly Event[] {
    if (this.#flushed) this.#lifecycle("stream flushed twice");
    if (!this.#finishSeen)
      this.#lifecycle("stream ended without finish_reason");
    this.#flushed = true;
    const events: Event[] = [];
    if (this.#thinkingOpen) {
      events.push({ type: "content_block_stop", index: this.#thinkingIndex });
      this.#thinkingOpen = false;
    }
    if (this.#textOpen) {
      events.push({ type: "content_block_stop", index: this.#textIndex });
      this.#textOpen = false;
    }
    let index = this.#nextIrIndex;
    for (const call of this.#toolCalls) {
      if (call.skipped) continue;
      if (call.id === undefined)
        this.#lifecycle(`tool_calls[${call.nativeIndex}] is missing final ID`);
      if (call.name === "")
        this.#lifecycle(
          `tool_calls[${call.nativeIndex}] is missing final function name`,
        );
      events.push({
        type: "content_block_start",
        index,
        block: {
          type: "tool_use",
          id: call.id,
          name: call.name,
          input: jsonText(call.fragments.join("")),
        },
      });
      for (const fragment of call.fragments)
        events.push({
          type: "content_block_delta",
          index,
          delta: { type: "input_json_delta", partial_json: jsonText(fragment) },
        });
      events.push({ type: "content_block_stop", index });
      index += 1;
      this.#nextIrIndex = index;
    }
    events.push(
      {
        type: "message_delta",
        stop_reason: this.#stopReason,
        usage: this.#usage,
      },
      { type: "message_done" },
    );
    return events;
  }

  Losses(): readonly Loss[] {
    return this.#losses;
  }

  #recordToolCalls(calls: readonly ChatCompletionsToolCallDelta[]): void {
    for (const delta of calls) {
      if (
        !Number.isInteger(delta.index) ||
        delta.index < 0 ||
        delta.index > this.#toolCalls.length
      )
        this.#lifecycle(
          `tool_calls index ${delta.index} is not the next consecutive native index`,
        );
      let call = this.#toolCalls[delta.index];
      if (call === undefined) {
        call = {
          nativeIndex: delta.index,
          id: undefined,
          name: "",
          fragments: [],
          skipped: false,
        };
        this.#toolCalls.push(call);
      }
      if (delta.id !== undefined && delta.id !== "") {
        if (call.id !== undefined && call.id !== delta.id)
          this.#lifecycle(`tool_calls[${delta.index}] has conflicting IDs`);
        call.id = delta.id;
      }
      if (
        delta.type !== undefined &&
        delta.type !== "function" &&
        !call.skipped
      ) {
        this.#losses.push({
          path: `choices[0].delta.tool_calls[${delta.index}]`,
          field: "type",
          reason: "unsupported-semantic",
          detail: `Chat Completions streamed tool type ${JSON.stringify(delta.type)} has no IR equivalent`,
        });
        call.skipped = true;
      }
      if (call.skipped || delta.function === undefined) continue;
      if (delta.function.name !== undefined) call.name += delta.function.name;
      if (delta.function.arguments !== undefined)
        call.fragments.push(delta.function.arguments);
    }
  }

  #finish(reason: string): StopReason {
    switch (reason) {
      case "stop":
        return "end_turn";
      case "length":
        return "max_tokens";
      case "content_filter":
        return "refusal";
      case "tool_calls":
        return "tool_use";
      default:
        this.#losses.push({
          path: "choices[0]",
          field: "finish_reason",
          reason: "unmapped-value",
          detail: `Chat Completions finish_reason ${JSON.stringify(reason)} maps to other`,
        });
        return "other";
    }
  }
  #lifecycle(message: string): never {
    throw new OxaError("stream-lifecycle", `chatcompletions: ${message}`);
  }
}

export interface ChatCompletionsStreamEncoderOptions {
  readonly modelMapper?: ModelMapper;
}

type EncoderBlock =
  | { readonly kind: "text"; readonly index: number }
  | {
      readonly kind: "thinking";
      readonly index: number;
      sawSignatureDelta: boolean;
    }
  | {
      readonly kind: "tool";
      readonly index: number;
      readonly id: string;
      readonly name: string;
      readonly input: string;
      readonly nativeIndex: number;
      readonly fragments: string[];
      started: boolean;
    };

/** Incrementally converts IR stream events to Chat Completions chunks. */
export class ChatCompletionsStreamEncoder {
  readonly #modelMapper: ModelMapper | undefined;
  #id = "";
  #model = "";
  #started = false;
  #finished = false;
  #done = false;
  #active: EncoderBlock | undefined;
  #nextIrIndex = 0;
  #nextNativeTool = 0;
  #toolSeen = false;
  #orderingDegrade = false;
  #pendingTools: ChatCompletionsChunk[] = [];

  constructor(options: ChatCompletionsStreamEncoderOptions = {}) {
    this.#modelMapper = options.modelMapper;
  }

  Apply(event: Event): ConversionResult<readonly ChatCompletionsChunk[]> {
    if (this.#done || (this.#finished && event.type !== "message_done"))
      this.#lifecycle("event applied after stream termination");
    switch (event.type) {
      case "message_start":
        if (this.#started) this.#lifecycle("duplicate message_start");
        this.#started = true;
        this.#id = event.id;
        this.#model = mapModel(this.#modelMapper, event.model);
        return this.#result([this.#chunk({ role: "assistant" })]);
      case "content_block_start":
        return this.#startBlock(event);
      case "content_block_delta":
        return this.#delta(event);
      case "content_block_stop":
        return this.#stopBlock(event.index);
      case "message_delta":
        return this.#terminal(event.stop_reason, event.usage);
      case "message_done":
        if (!this.#finished)
          this.#lifecycle("message_done out of grammar order");
        this.#done = true;
        return this.#result([]);
    }
  }

  #startBlock(
    event: Extract<Event, { type: "content_block_start" }>,
  ): ConversionResult<readonly ChatCompletionsChunk[]> {
    if (!this.#started || this.#active !== undefined)
      this.#lifecycle("content_block_start out of grammar order");
    if (event.index !== this.#nextIrIndex)
      this.#lifecycle(
        `content_block_start index ${event.index}, want ${this.#nextIrIndex}`,
      );
    this.#nextIrIndex += 1;
    if (event.block.type === "text") {
      if (this.#toolSeen) this.#orderingDegrade = true;
      this.#active = { kind: "text", index: event.index };
      return this.#result([]);
    }
    if (event.block.type === "thinking") {
      this.#active = {
        kind: "thinking",
        index: event.index,
        sawSignatureDelta: false,
      };
      if (event.block.signature === undefined) return this.#result([]);
      return {
        value: [],
        losses: [
          {
            path: `events[${event.index}].signature`,
            field: "signature",
            reason: "unmapped-field",
            detail:
              "Chat Completions reasoning_content has no signature field; the opaque ThinkingBlock signature is lost",
          },
        ],
      };
    }
    if (event.block.type !== "tool_use")
      this.#lifecycle(`unsupported content block ${event.block.type}`);
    if (event.block.id === "" || event.block.name === "")
      this.#lifecycle("tool_use requires nonempty id and name");
    this.#active = {
      kind: "tool",
      index: event.index,
      id: event.block.id,
      name: event.block.name,
      input: event.block.input,
      nativeIndex: this.#nextNativeTool++,
      fragments: [],
      started: false,
    };
    this.#toolSeen = true;
    return this.#result([]);
  }

  #delta(
    event: Extract<Event, { type: "content_block_delta" }>,
  ): ConversionResult<readonly ChatCompletionsChunk[]> {
    if (this.#active === undefined || event.index !== this.#active.index)
      this.#lifecycle("content_block_delta out of grammar order");
    if (this.#active.kind === "text") {
      if (event.delta.type !== "text_delta")
        this.#lifecycle("text block received non-text delta");
      return this.#result([this.#chunk({ content: event.delta.text })]);
    }
    if (this.#active.kind === "thinking") {
      if (event.delta.type === "thinking_delta") {
        if (this.#active.sawSignatureDelta)
          this.#lifecycle("thinking_delta after signature_delta");
        return this.#result([
          this.#chunk({ reasoning_content: event.delta.text }),
        ]);
      }
      if (event.delta.type === "signature_delta") {
        if (this.#active.sawSignatureDelta)
          this.#lifecycle("duplicate signature_delta");
        this.#active.sawSignatureDelta = true;
        return {
          value: [],
          losses: [
            {
              path: `events[${event.index}].signature`,
              field: "signature",
              reason: "unmapped-field",
              detail:
                "Chat Completions reasoning_content has no signature field; the opaque signature delta is lost",
            },
          ],
        };
      }
      this.#lifecycle("thinking block received non-thinking delta");
    }
    if (event.delta.type !== "input_json_delta")
      this.#lifecycle("tool block received non-input-json delta");
    const fragment = event.delta.partial_json as string;
    this.#active.fragments.push(fragment);
    this.#queueToolArguments(this.#active, fragment);
    return this.#result([]);
  }

  #stopBlock(index: number): ConversionResult<readonly ChatCompletionsChunk[]> {
    if (this.#active === undefined || index !== this.#active.index)
      this.#lifecycle("content_block_stop out of grammar order");
    if (this.#active.kind === "tool") {
      if (this.#active.fragments.length === 0) {
        this.#active.fragments.push(this.#active.input);
        this.#queueToolArguments(this.#active, this.#active.input);
      }
      if (this.#active.fragments.join("") !== this.#active.input)
        this.#lifecycle(
          "tool_use input does not equal concatenated input_json_delta fragments",
        );
    }
    this.#active = undefined;
    return this.#result([]);
  }

  #terminal(
    stopReason: StopReason,
    usage: Usage,
  ): ConversionResult<readonly ChatCompletionsChunk[]> {
    if (!this.#started || this.#active !== undefined)
      this.#lifecycle("message_delta out of grammar order");
    const inputTokens = encodeUsageInteger(
      usage.input_tokens,
      "usage.input_tokens",
    );
    const outputTokens = encodeUsageInteger(
      usage.output_tokens,
      "usage.output_tokens",
    );
    this.#finished = true;
    const losses: Loss[] = [];
    const finishReason = this.#finishReason(stopReason, losses);
    if (this.#orderingDegrade)
      losses.push({
        path: "events",
        field: "ordering",
        reason: "degraded",
        detail:
          "N-S-10: the text block after a tool block is normalized ahead of the tool calls; IR source order is not preserved",
      });
    const chunks = this.#pendingTools;
    this.#pendingTools = [];
    chunks.push({
      id: this.#id,
      object: "chat.completion.chunk",
      created: 0,
      model: this.#model,
      choices: [{ index: 0, delta: {}, finish_reason: finishReason }],
      usage: {
        prompt_tokens: inputTokens,
        completion_tokens: outputTokens,
        total_tokens: inputTokens + outputTokens,
        ...(usage.input_tokens_details === undefined
          ? {}
          : {
              prompt_tokens_details: {
                cached_tokens: encodeUsageInteger(
                  usage.input_tokens_details.cached_tokens,
                  "usage.input_tokens_details.cached_tokens",
                ),
              },
            }),
        ...(usage.output_tokens_details === undefined
          ? {}
          : {
              completion_tokens_details: {
                reasoning_tokens: encodeUsageInteger(
                  usage.output_tokens_details.reasoning_tokens,
                  "usage.output_tokens_details.reasoning_tokens",
                ),
              },
            }),
      },
    });
    return { value: chunks, losses };
  }

  #queueToolArguments(
    block: Extract<EncoderBlock, { kind: "tool" }>,
    arguments_: string,
  ): void {
    const toolCall: ChatCompletionsToolCallDelta = !block.started
      ? {
          index: block.nativeIndex,
          id: block.id,
          type: "function",
          function: { name: block.name, arguments: arguments_ },
        }
      : { index: block.nativeIndex, function: { arguments: arguments_ } };
    block.started = true;
    this.#pendingTools.push(this.#chunk({ tool_calls: [toolCall] }));
  }

  #chunk(delta: ChatCompletionsDelta): ChatCompletionsChunk {
    return {
      id: this.#id,
      object: "chat.completion.chunk",
      created: 0,
      model: this.#model,
      choices: [{ index: 0, delta, finish_reason: null }],
    };
  }

  #result(
    value: readonly ChatCompletionsChunk[],
  ): ConversionResult<readonly ChatCompletionsChunk[]> {
    return { value, losses: [] };
  }

  #finishReason(stopReason: StopReason, losses: Loss[]): string {
    switch (stopReason) {
      case "end_turn":
        return "stop";
      case "max_tokens":
        return "length";
      case "refusal":
        return "content_filter";
      case "tool_use":
        return "tool_calls";
      case "stop_sequence":
        losses.push({
          path: "",
          field: "stop_sequence",
          reason: "unmapped-value",
          detail:
            'Chat Completions finish_reason "stop" does not identify the matched stop sequence',
        });
        return "stop";
      default:
        this.#lifecycle(
          `stop reason ${JSON.stringify(stopReason)} has no Chat Completions equivalent`,
        );
    }
  }

  #lifecycle(message: string): never {
    throw new OxaError("stream-lifecycle", `chatcompletions: ${message}`);
  }
}
